### PR TITLE
Rework the aspect ratio calculation (Fixes 240p mode)

### PR DESCRIPTION
**Pull request by phire**
Calculations are now based on signal timings rather than pixels, as it didn't make a lot of sense to do things with pixels.

Now handles all 240i/240p/480i/480p modes without any special casing. Will even handle modes like 960i or 243@59.74hz

Despite the changes in equations, the resulting aspect ratios are identical.

### CODE REVIEW

**Comment by phire** [comment]
@mirrorbender could you take a look at this? It should have the exact same aspect ratios as your old code.

**Comment by phire** [comment]
This is ready for review/merging now. Replaces #3346

**Comment by JMC47** [comment]
lgtm in basic testing of various games and aspect ratios.

File: Source/Core/Core/HW/VideoInterface.cpp [hunk 1]
```diff
@@ -472,55 +472,84 @@ static u32 GetTicksPerOddField()
...
+		horizontal_active_ratio *= 63.555555f / 52.655555f;
+		// Even 486 active lines isn't completely agreed upon.
+		// Depending on how you count the two half lines you could get 485 or 484
+		vertical_active_ratio *= 525.0f / 486.0f;
 	}
-	return ((float)width / (float)height) * pixelAR;
+
+	// 5. Calculate the final ratio and scale to 4:3
+	float ratio = horizontal_active_ratio / vertical_active_ratio;
+	if(std::isnormal(ratio))	// Check we have a sane ratio and haven't propagated any infs/nans/zeros
```
Comment (delroth) [inline Source/Core/Core/HW/VideoInterface.cpp]: Space after if.

File: Source/Core/Core/HW/VideoInterface.cpp [hunk 2]
```diff
@@ -472,55 +472,84 @@ static u32 GetTicksPerOddField()
...
+	float vertical_period = tick_length * GetTicksPerField();
+	float horizontal_period= tick_length * GetTicksPerHalfLine() * 2;
+	float vertical_active_area = active_lines * horizontal_period;
+	float horizontal_active_area = tick_length * GetTicksPerSample() * active_width_samples;
+
+	// We are approximating the horizontal/vertical flyback transformers that control the
+	// position of the election beam on the screen. Our flyback transformers create a
+	// perfect Sawtooth wave, with a smooth rise and a fall that takes zero time.
+	// For more accurate emulation of video signals out of the 525 or 625 line standards,
+	// it might be necessary to emulate a less flyback transformer with more flaws.
```
Comment (delroth) [inline Source/Core/Core/HW/VideoInterface.cpp]: Missing the word "accurate" here.

**Comment by delroth** [comment]
Other than these two nits, code LGTM. @dolphin-emu-bot allowmerge

**Comment by mirrorbender** [comment]
@phire I looked over it quickly.  The numbers are correct (as much as that is possible here).  I can go over the code more closely if you want me to, but I trust that you and the others who have reviewed and tested it have already done a fine job of that, and at a glance it looks fine to me.  So I'd say go ahead and merge it.

**Comment by phire** [comment]
@mirrorbender Now that I've separated the vertical and horizontal calculations and moved away from pixels, I can see that the vertical timings are wrong by half a line per field. The NTSC timings should be `262.5 / 242.5` which works out to `525 / 485` (instead of the current `525 / 486`)

You can confirm this by looking Figure 7 on page 11 of [SMPTE 170M-2004](http://www.dxdlw.com/bbsupfile/2012%5C02%5C10%5C1734201163%5CSMPTE%20170M-2004%20Composite%20Analog%20Video%20Signal--NTSC%20for%20Studio%20Applications.pdf).
Actually it goes a step further and defines the vertical blanking per field as 20 lines + 1.5us in `Detail Y-Y` and Table 3 on page 13.

Half a line makes no sense in pixels, but it makes perfect sense for a CRT where the horizontal and vertical timings aren't directly connected and the image is rotated slightly anyway.

I'll submit a second PR in the future to address this.

**Comment by mirrorbender** [comment]
I'm having trouble following the diagrams in terms of where the half lines go in the picture (half line at the end of field 1 and beginning of field 2?), but my reasoning for 486 was that two fields of 242.5 resulted in 484 full lines plus 2 separate half lines, which meant 486 was a better number if those half-lines were scanned to it's own vertical space within the active frame.  However, assuming a slight tilt, I can see how 485 lines would be a better approximation of the appropriate height of the screen in a CRT TV.

**Comment by phire** [comment]
The half lines are at the start and end of the frame. The standard doesn't say it, but Field 1 is the even/lower field and Field 2 is the upper/odd field. 
With the tilt, the start of line 1 is the same height on the screen as the start of line 2, except line 1 starts half-way along the screen. It's the scan pattern which is tilted, I suspect TVs were build with a slight rotation of the coils in the opposite direction, so the tilt would be canceled out.

For a digital video format which can't do half-lines, padding it to 486 lines (or discarding both half-lines and dropping down to 484 lines) make sense, but 485 lines is completely nonsensical.



Interlaced basically means the vertical period and horozontal period are out of sync by half a line, so when the top of the upper field starts, it's half a

**Comment by phire** [comment]
Comments have been updated to fix typos and add some details from the BT-470-6 standard.

**Comment by fallaha56** [comment]
hi, aspect ratios still broken on N64 VC games -> going 16:9 when 4:3, bug report filed

**Comment by phire** [comment]
Technically that's correct emulation, as on a real Wii/TV you are required
to manually switch between 16:9 and 4:3 modes for virtual console games.

But we really should add .ini files for all virtual console games to
override that.

____________
Scott Mansell

**Comment by JMC47** [comment]
Yes, any issue made is invalid.

**Comment by fallaha56** [comment]
thanks @phire i did wonder if you guys were simply being too accurate ;-)

but yeah some form of override would be cool

@JMC47 wouldn't describe it as invalid! it makes using VC a massive pain...

**Comment by JMC47** [comment]
Hehehe!  It is accurate to the Wii, though.  That's why it's not valid.

**Comment by JMC47** [comment]
I am okay with ini overrides, though.

**Comment by fallaha56** [comment]
thanks @JMC47...fingers hurting from the added mouseclicks resetting ratios ;) can refile issue as 'finger injury' if that's better lol ;))

**Pull request merged**